File: Data Structures & Algorithms/depth-of-binary-tree/submission-10.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        max_depth = 1
        s = [(root, max_depth)]

        while s:
            node, depth = s.pop()

            max_depth = max(max_depth, depth)

            if node.left:
                s.append((node.left, depth + 1))
            if node.right:
                s.append((node.right, depth + 1))
        
        return max_depth

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-10.swift
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     var val: Int
 *     var left: TreeNode?
 *     var right: TreeNode?
 *     init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */

class Solution {
    func rightSideView(_ root: TreeNode?) -> [Int] {
        guard let root else { return [] }

        var res: [Int] = []
        var q: [TreeNode] = [root]
        var idx = 0
        
        while idx < q.count {
            let size = q.count - idx
            var rightmostNode: TreeNode?
            for _ in 1...size {
                let node = q[idx]
                idx += 1

                rightmostNode = node

                if let left = node.left { q.append(left) }
                if let right = node.right { q.append(right) }
            }
            if let rightmostNode { res.append(rightmostNode.val) }
        }

        return res
    }
}
